### ISSUE TITLE
Implement USB DbC

### ISSUE THREAD

**Issue by alice-sowerby** [FreeBSDFoundation/proj-laptop#38]
## User Story
As a laptop developer, I want to be able to debug while developing suspend/resume via a USB-connected machine.

## Implementation Notes
The work as POC has been created, but needs making ready for production, testing etc. This includes getting the source code into the right format etc. 

## Acceptance Criteria

- [x] It's possible to run a debugger running on a host machine which can run one (must) or both (should) LLDB or GDB debugging on the target laptop.
- [x] Target laptop will be the main one from the list of target laptops #24 
- [x] USB-A (must)
- [x] USB-C (should)
## Discussion thread
https://lists.freebsd.org/archives/freebsd-desktop/2025-March/005616.html

**Comment by mhorne** [comment]
Here is my update, overdue, on this item.

First, some background. This functionality was implemented by `hrs@FreeBSD.org`, and presented at BSDCan 2024 (at least). See the [slides](https://people.allbsd.org/~hrs/FreeBSD/udbc/sato-20240531-usbdbc.pdf).

He has made several versions available publicly. Relevant to my testing so far:
 - [20240531](https://people.allbsd.org/~hrs/FreeBSD/udbc/20240531/)/BSDCan version: sources as tarball
 - [20240721](https://people.allbsd.org/~hrs/FreeBSD/udbc/20240721/): sources as tarball
 - [20240809](https://people.allbsd.org/~hrs/FreeBSD/udbc/20240809/): memstick `.img` only.

First, I took the **20240531** version of the sources, and rebased it onto FreeBSD's `main` branch. With this version, I had limited functional success:
 - Initiation of the DbC connection possible from `loader(8)`, debug host can connect to the console using `udbc` driver and receive characters sent by debug target.
 - DbC reported as unsupported by the kernel xhci driver; no DbC connection possible from this context (see snippet). 

```
xhci0: <XHCI (generic) USB 3.0 controller> mem 0xef720000-0xef72ffff at device 20.0 on pci0
xhci0: 32 bytes context size, 64-bit DMA
xhci_debug_probe: looking for 02
xhci_debug_probe: looking for 02
xhci_debug_probe: looking for c0
xhci_debug_probe: looking for 01
xhci_debug_probe: looking for c6
xhci_debug_probe: looking for c7
xhci_debug_probe: looking for c2
xhci_debug_probe: looking for c3
xhci_debug_probe: looking for c4
xhci_debug_probe: looking for c5
xhci_debug_probe: looking for c8
xhci_debug_probe: looking for c9
xhci_debug_probe: looking for ca
xhci_debug_probe: looking for cb
xhci_debug_probe: DBC_ID was not found
xhci0: DbC not available
usbus0 on xhci0
usbus0: 5.0Gbps Super Speed USB v3.0
```

Despite the claim, my laptop hardware does indeed support the debug capability, as proven by the successful DbC connection from `loader(8)`. Windows also reports that all external/physical USB ports have this capability. Therefore, there is a bug in the feature detection added to the `xhci` driver.

So, to proceed I applied the updated sources from the **20240721** version. This refactored the code substantially. The results were less usable:
 - loader(8) presents a blank screen and cannot be used (unless `udbc_console` is disabled, defeating the purpose).
 - Kernel xhci driver fails in the same way; `DbC not available`.

My git version of this in-progress code, rebased on recent `main` is available here: https://github.com/mhorne/freebsd/tree/udbc

See also #48.

Finally, I have experimented with the **20240809** version, distributed as a bootable FreeBSD image/installer, but not as source code. This version, and its usage, are described in detail on the [wiki page](https://wiki.freebsd.org/HirokiSato/xhci).

Based on the console messages only, it is clear there is some significant code changes between the **20240721** and **20240809** versions. Running this on my debug target (laptop), I got the following results:
 - `loader(8)` is functional. The `udb_probe` command emits the message: `No DbC cable detected`.
 - The kernel `xhci` driver successfully detects the DbC capability, but the `udbcons` sub-device fails to attach (see below snippet)

```
xhci0: <XHCI (generic) USB 3.0 controller> mem 0xef720000-0xef72ffff at device 20.0 on pci0
xhci0: 32 bytes context size, 64-bit DMA
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0020
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0050
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c0, next=03fc
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=01, next=0088
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c6, next=000c
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c7, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c2, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=0a, next=0040
xhci_debug_get_xecp: DbC was found at 00008700
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0020
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0050
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c0, next=03fc
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=01, next=0088
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c6, next=000c
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c7, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c2, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=0a, next=0040
xhci_debug_get_xecp: DbC was found at 00008700
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0020
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0050
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c0, next=03fc
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=01, next=0088
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c6, next=000c
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c7, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c2, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=0a, next=0040
xhci_debug_get_xecp: DbC was found at 00008700
udbcons0: <USB xHCI DbC Console> on xhci0
udbcons0: Fixing up RID: 0x2b00
udbcons0: RID not found: expected=0x00a0
device_attach: udbcons0 attach returned 6
usbus0 on xhci0
usbus0: 5.0Gbps Super Speed USB v3.0
```

So now there is some trouble with the resource allocation/management. (I am not quite sure what is being attempted when it says `Fixing up RID: 0x2b00`)

Stranger still, it succeeds for the second `xhci` root hub, `xhci1`:
```
xhci1: <Intel Alpine Ridge Thunderbolt 3 USB controller> mem 0xe0000000-0xe000ffff at device 0.0 on pci5
xhci1: 32 bytes context size, 64-bit DMA
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0070
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c0, next=03fc
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=01, next=0088
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c6, next=000c
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c7, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c2, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=0a, next=0040
xhci_debug_get_xecp: DbC was found at 00008700
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0070
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c0, next=03fc
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=01, next=0088
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c6, next=000c
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c7, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c2, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=0a, next=0040
xhci_debug_get_xecp: DbC was found at 00008700
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=02, next=0070
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c0, next=03fc
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=01, next=0088
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c6, next=000c
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c7, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=c2, next=0100
xhci_debug_get_xecp: Looking for xECP: expected=0a, found=0a, next=0040
xhci_debug_get_xecp: DbC was found at 00008700
udbcons1: <USB xHCI DbC Console> on xhci1
udbcons1: Fixing up RID: 0x2b00
udbcons1: DbC xECP found at 0x8700
udbcons1: Creating /dev/udbcons1
udbcons1: state: DISCONNECTED(0x10) -> OFF(0x00)
udbcons1: state: OFF(0x00) -> DISCONNECTED(0x10)
udbcons1: No DbC cable detected
usbus1 on xhci1
usbus1: 5.0Gbps Super Speed USB v3.0
```

However, all external ports on this machine are connected to root hub 0, that is `xhci0`. Therefore, this version is still unusable for me. Because it was not distributed as source code, I can't do much to debug or learn more about what is going wrong.

So, from my view, we are blocked. I recently sent a message to @hrs-allbsd requesting a newer version of the sources (ideally the latest, but the **20240809** version would be usable for me). Until I hear back about this, I am somewhat out of ideas on how to proceed.

The alternative would be to continue hacking on the **20240721** version, and attempt to resolve the issues there, but I note that this would be retreading work already done (bugs fixed) in the **20240809** version of the code. Since I am not the principle author of this work, I think this would not be time well spent.

Maybe @adventureloop or @bzfbd can weigh in on their experience or successes with this feature.

**Comment by alice-sowerby** [comment]
@adventureloop please could you add an update for the work you have been doing on this too? (And we co-assigned this issue to you with Mitchell.)

**Comment by adventureloop** [comment]
This branch has hacks which enable debugging from a physical host to a virtual machine guest target. 

https://github.com/mhorne/freebsd/pull/1

I don't have a supported laptop I can use as a target, without doing some radical changes to my development environment, but I don't see why it wouldn't be functional considering the xhci controller is the same on my framework 13 as my test target.

Next steps are to tidy up the branch and probably strip out the !amd64 support at least until we can get wider testing.

**Comment by mhorne** [comment]
Some meaningful progress has been made here. In short, we now have an updated version of the prototype, with several tweaks applied, and I have (more-or-less) verified/reproduced all functionality as originally advertised.

The code still requires some substantial changes/improvements before we can begin to consider merging. In particular, much effort was expended by the original author into providing the USB DbC console to FreeBSD's bootloader, `loader(8)`.  In turn, this has complicated the code, and presented some significant debugging barriers. Currently the kernel's `xhci(4)` driver depends on the loader's setup/inherited state in order to activate the DbC console, and I am working to decouple it. 

After some discussion with @adventureloop, we concluded that this loader portion, while interesting, likely shouldn't be part of the acceptance criteria for the upstreamed version, as it does not present as compelling of a debugging aid. It may be that we remove it completely.

My current version of the code, in a reasonably cleaned-up and functional state, is (again) available here:

https://github.com/mhorne/freebsd/tree/udbc

**Comment by alice-sowerby** [comment]
Backporting info from the June report:

Kernel debugging using the XHCI debug standard is working as a proof of concept. In June, work began to update the patch set to improve the interface between the loader and the kernel. These changes should make XHCI debug more robust and address occasional panics when booting a kernel. There remain issues with performance over the emulated serial link, and crashes can occur when using XHCI debug as a system console.

**Comment by adventureloop** [comment]
All pre commit issues have been resolved and we are now working on extracting logical changes and upstreaming

**Comment by adventureloop** [comment]
I have rebased the loader review and added the required kernel support in a child revision:
https://reviews.freebsd.org/D53472
https://reviews.freebsd.org/D56602

I have been trying to debug stalls in the host side udbc driver on freebsd.
